File: backend/app/schemas/chatbot.py
from pydantic import BaseModel, Field
from typing import Dict, List, Any, Optional
from datetime import datetime
from enum import Enum

class ConversationStatusEnum(str, Enum):
    ACTIVE = "active"
    ARCHIVED = "archived"
    DELETED = "deleted"

class MessageRoleEnum(str, Enum):
    USER = "user"
    ASSISTANT = "assistant"
    SYSTEM = "system"

class IntentTypeEnum(str, Enum):
    GENERAL_CHAT = "general_chat"
    INFRASTRUCTURE_QUERY = "infrastructure_query"
    TROUBLESHOOTING = "troubleshooting"
    RESOURCE_ANALYSIS = "resource_analysis"
    COST_OPTIMIZATION = "cost_optimization"
    MONITORING_ALERT = "monitoring_alert"
    REMEDIATION_REQUEST = "remediation_request"
    HELP_REQUEST = "help_request"

# Request/Response Schemas

class ConversationCreateRequest(BaseModel):
    title: Optional[str] = Field(None, description="Conversation title")
    context: Optional[Dict[str, Any]] = Field(None, description="Initial context")

class ConversationResponse(BaseModel):
    id: int
    session_id: str
    user_id: int
    title: Optional[str]
    status: ConversationStatusEnum
    context: Optional[Dict[str, Any]]
    total_messages: int
    total_tokens_used: int
    last_activity: Optional[datetime]
    created_at: datetime
    updated_at: Optional[datetime]

class ConversationListResponse(BaseModel):
    conversations: List[ConversationResponse]
    total: int
    page: int
    per_page: int

class MessageResponse(BaseModel):
    id: int
    conversation_id: int
    role: MessageRoleEnum
    content: str
    model_used: Optional[str]
    tokens_used: int
    response_time: Optional[float]
    cached: bool
    context_snapshot: Optional[Dict[str, Any]]
    created_at: datetime

class ConversationWithMessagesResponse(BaseModel):
    conversation: ConversationResponse
    messages: List[MessageResponse]

class EnhancedChatRequest(BaseModel):
    message: str = Field(..., min_length=1, max_length=4000, description="User message")
    session_id: Optional[str] = Field(None, description="Session ID for conversation context")
    context: Optional[Dict[str, Any]] = Field(default_factory=dict, description="Additional context")
    oci_context: Optional[Dict[str, Any]] = Field(None, description="OCI-specific context")
    enable_intent_recognition: bool = Field(True, description="Enable intent recognition")
    use_templates: bool = Field(True, description="Allow template suggestions")

class EnhancedChatResponse(BaseModel):
    response: str = Field(..., description="AI response")
    session_id: str = Field(..., description="Session ID")
    conversation_id: Optional[int] = Field(None, description="Database conversation ID")
    model: str = Field(..., description="Model used")
    tokens_used: int = Field(..., description="Tokens consumed")
    response_time: float = Field(..., description="Response time in seconds")
    cached: bool = Field(False, description="Whether response was cached")
    intent: Optional["IntentResponse"] = Field(None, description="Detected intent")
    suggested_templates: List["TemplateResponse"] = Field(default_factory=list, description="Suggested query templates")
    oci_insights: Optional[Dict[str, Any]] = Field(None, description="OCI-specific insights")

class IntentResponse(BaseModel):
    intent_type: IntentTypeEnum
    confidence_score: float
    entities: Optional[Dict[str, Any]]

class QueryTemplateRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=100, description="Template name")
    description: Optional[str] = Field(None, description="Template description")
    category: str = Field(..., description="Template category")
    template_text: str = Field(..., description="Template text with variables")
    variables: Optional[Dict[str, Any]] = Field(None, description="Variable definitions")
    requires_role: Optional[str] = Field(None, description="Minimum role required")

class TemplateResponse(BaseModel):
    id: int
    name: str
    description: Optional[str]
    category: str
    template_text: str
    variables: Optional[Dict[str, Any]]
    usage_count: int
    requires_role: Optional[str]
    is_active: bool
    created_at: datetime
    updated_at: Optional[datetime]

class TemplateListResponse(BaseModel):
    templates: List[TemplateResponse]
    total: int
    categories: List[str]

class UseTemplateRequest(BaseModel):
    template_id: int = Field(..., description="Template ID to use")
    variables: Optional[Dict[str, Any]] = Field(None, description="Variable values")
    session_id: Optional[str] = Field(None, description="Session ID")
    context: Optional[Dict[str, Any]] = Field(None, description="Additional context")

class ConversationExportRequest(BaseModel):
    session_id: str = Field(..., description="Session ID to export")
    format: str = Field("json", description="Export format (json, csv, markdown)")
    include_metadata: bool = Field(True, description="Include metadata in export")

class ConversationExportResponse(BaseModel):
    session_id: str
    format: str
    content: str
    filename: str
    exported_at: datetime

class ConversationAnalyticsResponse(BaseModel):
    user_id: int
    period: str
    total_conversations: int
    total_messages: int
    total_tokens: int
    avg_response_time: float
    intent_breakdown: Dict[str, int]
    top_queries: List[str]
    most_used_templates: List[Dict[str, Any]]

class ChatbotFeedbackRequest(BaseModel):
    session_id: str = Field(..., description="Session ID")
    message_id: Optional[int] = Field(None, description="Specific message ID")
    rating: int = Field(..., ge=1, le=5, description="Rating (1-5)")
    feedback_text: Optional[str] = Field(None, description="Feedback text")
    feedback_type: Optional[str] = Field(None, description="Feedback type")

class FeedbackResponse(BaseModel):
    id: int
    conversation_id: int
    message_id: int
    user_id: int
    rating: int
    feedback_text: Optional[str]
    feedback_type: Optional[str]
    created_at: datetime

class ChatbotHealthResponse(BaseModel):
    status: str
    database_connection: bool
    redis_connection: bool
    ai_service: bool
    intent_recognition: bool
    total_conversations: int
    active_sessions: int
    avg_response_time: float

class SuggestedQueriesResponse(BaseModel):
    infrastructure_queries: List[str]
    monitoring_queries: List[str]
    troubleshooting_queries: List[str]
    cost_queries: List[str]

# Update forward references
EnhancedChatResponse.model_rebuild() 

File: frontend/src/types/chatbot.ts
// Enums
export enum ConversationStatus {
  ACTIVE = "active",
  ARCHIVED = "archived",
  DELETED = "deleted"
}

export enum MessageRole {
  USER = "user",
  ASSISTANT = "assistant",
  SYSTEM = "system"
}

export enum IntentType {
  GENERAL_CHAT = "general_chat",
  INFRASTRUCTURE_QUERY = "infrastructure_query",
  TROUBLESHOOTING = "troubleshooting",
  RESOURCE_ANALYSIS = "resource_analysis",
  COST_OPTIMIZATION = "cost_optimization",
  MONITORING_ALERT = "monitoring_alert",
  REMEDIATION_REQUEST = "remediation_request",
  HELP_REQUEST = "help_request"
}

// Request Types
export interface EnhancedChatRequest {
  message: string;
  session_id?: string;
  context?: Record<string, any>;
  oci_context?: Record<string, any>;
  enable_intent_recognition?: boolean;
  use_templates?: boolean;
}

export interface ConversationCreateRequest {
  title?: string;
  context?: Record<string, any>;
}

export interface UseTemplateRequest {
  template_id: number;
  variables?: Record<string, any>;
  session_id?: string;
  context?: Record<string, any>;
}

export interface ConversationExportRequest {
  session_id: string;
  format?: 'json' | 'csv' | 'markdown';
  include_metadata?: boolean;
}

export interface ChatbotFeedbackRequest {
  session_id: string;
  message_id?: number;
  rating: number; // 1-5 scale
  feedback_text?: string;
  feedback_type?: string;
}

// Response Types
export interface IntentResponse {
  intent_type: IntentType;
  confidence_score: number;
  entities?: Record<string, any>;
}

export interface TemplateResponse {
  id: number;
  name: string;
  description?: string;
  category: string;
  template_text: string;
  variables?: Record<string, any>;
  usage_count: number;
  requires_role?: string;
  is_active: boolean;
  created_at: string;
  updated_at?: string;
}

export interface EnhancedChatResponse {
  response: string;
  session_id: string;
  conversation_id?: number;
  model: string;
  tokens_used: number;
  response_time: number;
  cached: boolean;
  intent?: IntentResponse;
  suggested_templates: TemplateResponse[];
  oci_insights?: Record<string, any>;
}

export interface MessageResponse {
  id: number;
  conversation_id: number;
  role: MessageRole;
  content: string;
  model_used?: string;
  tokens_used: number;
  response_time?: number;
  cached: boolean;
  context_snapshot?: Record<string, any>;
  created_at: string;
}

export interface ConversationResponse {
  id: number;
  session_id: string;
  user_id: number;
  title?: string;
  status: ConversationStatus;
  context?: Record<string, any>;
  total_messages: number;
  total_tokens_used: number;
  last_activity?: string;
  created_at: string;
  updated_at?: string;
}

export interface ConversationListResponse {
  conversations: ConversationResponse[];
  total: number;
  page: number;
  per_page: number;
}

export interface ConversationWithMessagesResponse {
  conversation: ConversationResponse;
  messages: MessageResponse[];
}

export interface TemplateListResponse {
  templates: TemplateResponse[];
  total: number;
  categories: string[];
}

export interface ConversationExportResponse {
  session_id: string;
  format: string;
  content: string;
  filename: string;
  exported_at: string;
}

export interface ConversationAnalyticsResponse {
  user_id: number;
  period: string;
  total_conversations: number;
  total_messages: number;
  total_tokens: number;
  avg_response_time: number;
  intent_breakdown: Record<string, number>;
  top_queries: string[];
  most_used_templates: Array<{name: string; usage_count: number}>;
}

export interface FeedbackResponse {
  id: number;
  conversation_id: number;
  message_id: number;
  user_id: number;
  rating: number;
  feedback_text?: string;
  feedback_type?: string;
  created_at: string;
}

export interface ChatbotHealthResponse {
  status: string;
  database_connection: boolean;
  redis_connection: boolean;
  ai_service: boolean;
  intent_recognition: boolean;
  total_conversations: number;
  active_sessions: number;
  avg_response_time: number;
}

export interface SuggestedQueriesResponse {
  infrastructure_queries: string[];
  monitoring_queries: string[];
  troubleshooting_queries: string[];
  cost_queries: string[];
}

// UI-specific types
export interface ChatMessage extends MessageResponse {
  isUser: boolean;
  timestamp: Date;
  isLoading?: boolean;
  intent?: IntentResponse;
  suggested_templates?: TemplateResponse[];
  oci_insights?: Record<string, any>;
}

export interface ConversationSummary {
  id: number;
  session_id: string;
  title?: string;
  last_message?: string;
  last_activity: Date;
  message_count: number;
  status: ConversationStatus;
}

export interface TemplateCategory {
  name: string;
  templates: TemplateResponse[];
  count: number;
}

export interface ChatInputProps {
  value: string;
  onChange: (value: string) => void;
  onSend: () => void;
  onTemplateSelect?: (template: TemplateResponse) => void;
  placeholder?: string;
  disabled?: boolean;
  isLoading?: boolean;
  suggestedQueries?: string[];
}

export interface QuickAction {
  id: string;
  label: string;
  query: string;
  icon?: string;
  category?: string;
}

// Error types
export interface ChatbotError {
  message: string;
  code?: string;
  details?: any;
} 